### DISCUSSION TITLE
Collection Literals – Possible to satisfy compile-time constant requirement?

### DISCUSSION THREAD

**Discussion by ero-qt** [dotnet/csharplang#7484] [Language Ideas] [answered]
Would it be possible to allow collection literals to be passed, for example, as default parameters?

Something like this:
```cs
void M(int[] foo = [ ]) { }
```
today throws the compiler error `Default parameter value for 'foo' must be a compile-time constant`.

It would be fantastic to see this working one day.

**Comment by huoyaoyuan** [comment, marked as answer]
The limitation is from compiled assembly encoding, not the compile-time constant check. Currently, no collection type can be encoded in constants.

However for attributes we can encode array. Collection expressions may be permitted there.

**Comment by CyrusNajmabadi** [comment]
There is no way that that can be constant. Consider:

```
Int[] M(int[] foo = [ ]) {
   Foo[0] = 1;
   Return foo;
```

If this was constant, you'd now be mutating the constant value.

**Reply by jnm2** [reply]
Empty arrays are immutable, so there would have to be something inside the `[]`.

**Reply by HaloFour** [reply]
Would kinda be nice to be able to have a default value of an empty array.  C# could encode that as a custom attribute as it does with other pseudo-constants.

**Comment by TonyValenti** [comment]
@just-ero - for situations like that, you might consider using a functor approach or an args object with defaults on it.

**Comment by jnm2** [comment]
What are the motivating scenarios, things that get easier to do as a result?

**Reply by ero-qt** [reply]
Optional array (or other collections) parameters like shown in the example in the original post.

**Reply by jnm2** [reply]
Having concrete scenarios helps get people interested.

**Reply by michael-hawker** [reply]
This would be really nice to have for default initialization of array types. For instance, If I have a record today:

```cs
public record DataToProcess(string Id,
                            string Description = "", 
                            bool? ProcessInParallel = true, 
                            FileInfo[]? DataFiles = null)
```

I'd much rather have:

```cs
public record DataToProcess(string Id,
                            string Description = "", 
                            bool? ProcessInParallel = true, 
                            FileInfo[] DataFiles = []) // Nice to have empty collection as default for later processing over null
```

Where `DataFiles` is initialized to an empty collection by default. As in the case it's not provided above (for whatever reason, there's more types of data in my scenario, but simplifying the example. i.e. I can't make one required as I may have others to process but won't know which, so pipeline just loops over each one), I want to be able to just loop over the empty array without a check first for null.

And in the case where data is provided then the default value is discarded anyway. It just makes null propagation and overhead much more taxing vs. using the better patten of empty collections as a default. i.e. I want to be able to just do this later:

```cs
foreach (var fle in dataToProcess.DataFiles) // Nice and easy! No overhead, not error prone/propagating
{
    // Do Work
}
```

Instead I have to do extra checks:
```cs
if (dataToProcess.DataFiles is not null) // Extra check vs. nice default value that's loopable
{
    foreach (var fle in dataToProcess.DataFiles)
    {
        // Do Work
    }
}
```

or be fancy at every loop site:

```cs
    foreach (var fle in dataToProcess.DataFiles ?? []) // Extra stuff everywhere else vs. initialization
    {
        // Do Work
    }
```

**Reply by martin-steinhoefer** [reply]
If I remember correctly, there was a guideline to avoid optional parameters in favor of overloads. I guess that was an FxCop rule (CA1026?). 
Those days, I did a little search on the internet and found some interesting facts (Phil Haack: [1](https://haacked.com/archive/2010/08/10/versioning-issues-with-optional-arguments.aspx/), [2](https://haacked.com/archive/2010/08/12/more-optional-versioning-fun.aspx/)). 
I think that those arguments still are to be considered when using default parameters.

**Reply by michael-hawker** [reply]
@martin-steinhoefer thanks for some context and resources, good reads. However, in these cases I'm not defining a public API surface I expect to version or have others consume. I just have an internal record class to store and hold data for processing, I'm in control of the surface, I just may not have all data at all times.

In general, it's best to initialize arrays/lists as empty over null, so I'd just like to make facilitating that easier, especially now with primary constructors.

(Some references to various guidance on empty array over null: [1](https://wiki.sei.cmu.edu/confluence/display/c/MSC19-C.+For+functions+that+return+an+array%2C+prefer+returning+an+empty+array+over+a+null+value), [2](https://learn.microsoft.com/en-us/dotnet/standard/design-guidelines/guidelines-for-collections), [3](https://thefinestartist.com/effective-java/43))

**Comment by MitchRazga** [comment]
Does anyone know if collection expressions will be permitted within attributes in C#12?

I can't find any discussion within the LDM or working group notes but the analyzer IDE0300 ("Collection initialization can be simplified") suggests changing this:
```
[UnmanagedCallConv(CallConvs = new Type[] { typeof(CallConvCdecl) })]
```
to:
```
[UnmanagedCallConv(CallConvs = [typeof(CallConvCdecl)])]
```
which is currently invalid (CS0182).

**Reply by jnm2** [reply]
Support is coming soon, just not implemented yet. https://github.com/dotnet/roslyn/issues/69133 is the issue to follow.
